### DISCUSSION TITLE
Is SPDLOG_COMPILED_LIB required when using the non header-only version?

### DISCUSSION THREAD

**Discussion by deaklajos** [gabime/spdlog#3059] [General]
I would like to use the **non** header only version of spdlog to reduce my build times(mostly template instantiations of `spdlog::pattern_formatter::handle_flag_<>`).
Tried to use:
```cmake
target_link_libraries(${EXECUTABLE_NAME} PRIVATE spdlog::spdlog)
```
instead of 
```cmake
target_link_libraries(${EXECUTABLE_NAME} PRIVATE spdlog::spdlog_header_only)
```

It didn't do seemingly anything until i added:
```cmake
target_compile_definitions(${EXECUTABLE_NAME} PRIVATE SPDLOG_COMPILED_LIB)
```

Am I missing something?
From [INSTALL](https://github.com/gabime/spdlog/blob/v1.x/INSTALL) I assumed `SPDLOG_COMPILED_LIB` is 
only needed if I copy the .cpp files to my build tree(bit confusing).
Also as an example in [tests/CMakeLists.txt](https://github.com/gabime/spdlog/blob/v1.x/tests/CMakeLists.txt) I would assume(from my experince) that target `spdlog-utests` will use header only even if is linked with `spdlog::spdlog`.

**Comment by tt4g** [comment]
Yes. If `SPDLOG_COMPILED_LIB` is not defined, spdlog will behave as a header-only library.

> From [INSTALL](https://github.com/gabime/spdlog/blob/v1.x/INSTALL) I assumed `SPDLOG_COMPILED_LIB` is
> only needed if I copy the .cpp files to my build tree(bit confusing).

Not strictly true.
You need to import spdlog into your CMake project before `add_target_link_libraries()`.

There are several ways to do this.
If spdlog is installed on your machine, you can use `find_package()`.
This method is also used in `tests`.

https://github.com/gabime/spdlog/blob/8fed530bdffbe69304a330be2096c65793b722c2/tests/CMakeLists.txt#L6

You can also use `FetchContent` and `add_subdirectory()` to import spdlogs.
Please read the CMake documentation for their features.

**Reply by deaklajos** [reply]
> Yes. If SPDLOG_COMPILED_LIB is not defined, spdlog will behave as a header-only library.

Then shouldn't `spdlog-utests` compiled with `SPDLOG_COMPILED_LIB`?
And https://github.com/gabime/spdlog/blob/8fed530bdffbe69304a330be2096c65793b722c2/INSTALL#L14-L15
Be 
```cmake
add_executable(example example.cpp)
target_link_libraries(example spdlog::spdlog)
target_compile_definitions(example SPDLOG_COMPILED_LIB)
```

**Reply by tt4g** [reply]
No, that is not necessary.
Using `spdlog` found by `find_package()` automatically defines the macros defined in `CMakeLists.txt` in spdlog for the associated application.

CMake target `spdlog` (`spdlog::spdlog` is alias of `spdlog`) is a compiled library, so `SPDLOG_COMPILED_LIB` is defined in `example` target.

https://github.com/gabime/spdlog/blob/8fed530bdffbe69304a330be2096c65793b722c2/CMakeLists.txt#L173

https://github.com/gabime/spdlog/blob/8fed530bdffbe69304a330be2096c65793b722c2/CMakeLists.txt#L180

This is a CMake feature, so please learn more about CMake.

**Reply by deaklajos** [reply]
Thanks, I didn't know that it would propagate definitions like that.
The problem was on our side, for every translation unit we created object libraries and in the end we link them together into an executable and only that time we linked spdlog.
I guess we should be linking every dependency to those object libs as well.

**Reply by tt4g** [reply]
Yes, all projects that reference the spdlog header file should have the same macros defined as when spdlog was built.
